src/Symfony/Component/DependencyInjection/Tests/Dumper/PhpDumperTest.php

```diff
@@ -1611,6 +1611,29 @@ public function testClosure()
 
         $this->assertStringEqualsFile(self::$fixturesPath.'/php/closure.php', $dumper->dump());
     }
+
+    public function testServiceClosureWithArray()
+    {
+        $container = new ContainerBuilder();
+        $container->register('foo', ObjectWithLazyConfig::class)
+            ->setPublic('true')
+            ->setArguments([
+                '$config' => new ServiceClosureArgument([
+                    'name' => 'Foo',
```

Comment:
I would remove this test case: passing an array is not a supported use case

Reply:
Ok, I removed it. But I think that a closure that returns an array would also be beneficial. Is this something to consider? Maybe by introducing a `ClosureArgument`? What's the best way to get this supported?

Reply:
If you look at setValues, the support for array is very limited. In fact, the code expects to deal with an array contain only one reference

do you have a real world use case? otherwise better not do anything :)

Reply:
Yes, my real world use case is that I'm implementing [graphql-php](https://webonyx.github.io/graphql-php/schema-definition/) objects as services. They require array configuration and heavily use closures for lazyness. I was super happy to see that I could just pass a field config array as a `ServiceClosureArgument`. 

With `ClosureArgument` support, we could even further improve this. For example, the `graphql-php` library wants callables for resolvers. In a compiler pass I want to create closures with dynamic arguments. That are then dumped to the container like this:

```php
function($value, $user) : callable {
  $repository = ($containerRef->get()->privates['repository'] ??= new SomeRepository());
  $mailer = ($containerRef->get()->privates['mailer'] ??= new Mailer());
  return $someService->someMethod($repository, $mailer, $value, $user);
}
```

Now the closure can be passed to anything that is then able to call it. It doesn't have to bother about fetching the services, as that is now a compiler/container responsibilty.

This could also be used for controller actions, becoming callable, with services injected / prepared. It would remove the need for `ServiceValueResolver` and a separate ServiceLocator that keeps track of those services.

A DSL would look something like this:
```php
new ClosureArgument(new Expression('service("SomeService").someMethod(service("SomeRepository"), service("Mailer"), $closure.value, $closure.user)'), ['value', 'user']);
```

Reply:
@nicolas-grekas Maybe better to create a separate issue for this?

Reply:
Yes, if we should follow up :)

Reply:
Oke, created the issue:
- https://github.com/symfony/symfony/issues/49489